alf/utils/common.py

```diff
@@ -325,6 +325,85 @@ def forward(self):
             self._counter = 0
 
 
+@alf.configurable
+class PeriodicReset(nn.Module):
+    r"""Performs a periodic reset to model parameters.
+
+    For each weight :math:`m` in the model, reset (reinitialization) is done periodically
+    according to the specified schedule:
+
+    .. math::
+
+        m <-- re-initialize.
+
+    Note: 
+        1) we reinitialize Network parameters and always reset buffers. For non-Network
+        parameters, we record their initial value and reassign those values when reset.
+        2) for a ``Network`` instance, if it implements a member function ``reset_parameters``,
+        then this function will be used to reset the network parameter, which provides
+        the user more flexibilities to achieve customized reset behaviors (e.g. only reset
+        certain layers). If ``reset_parameters`` does not exist, then all the network
+        parameters will be reset. 
+        
+    Args:
+        models (Network | list[Network] | Parameter | list[Parameter] ): the
+            models or parameters that will be reset periodically according to schedule.
+        period: Step interval or scheduler at which the models or parameters will be reset.
+            If negative, no reset is done.
+        reset_buffers: if True, the torch.buffer instances will also be reset
+        post_processings: a list of callables to be applied after reset. For example,
+            initial parameter copy to the target critic in some RL algorithms.
+    """
+
+    def __init__(self,
+                 models,
+                 period: Union[int, Scheduler] = 1,
+                 reset_buffers: bool = True,
+                 post_processings=List[Callable]):
+        super().__init__()
+        models = as_list(models)
+        self._models = models
+        self._period = as_scheduler(period)
+        self._counter = 0
+        self._reset_buffers = reset_buffers
+        self._post_processings = post_processings
+        # record the inital values of torch.nn.Parameter instances in ``models``
+        self._init_param_values = {
+            id(p): p.data.clone()
+            for p in models if isinstance(p, torch.nn.Parameter)
+        }
+
+    def _copy_model_or_parameter(self, s, t):
+        if isinstance(t, nn.Parameter):
+            t.data.copy_(s)
+        else:
+            for ws, wt in zip(s.parameters(), t.parameters()):
+                wt.data.copy_(ws)
+            if self._reset_buffers:
+                for ws, wt in zip(s.buffers(), t.buffers()):
+                    wt.copy_(ws)
+
+    def forward(self):
+        self._counter += 1
+        period = self._period()
```

Comment:
The scheduler uses global training iter as the counter, so `period` here is scheduled according to how many training iterations. But `self._counter` is maintained according to num of updates. If one iter contains multiple updates, then the comparison below `self._counter >= period` is not correct.

Reply:
> The scheduler uses global training iter as the counter, so `period` here is scheduled according to how many training iterations. But `self._counter` is maintained according to num of updates. If one iter contains multiple updates, then the comparison below `self._counter >= period` is not correct.

This is a known behavior. Already discussed this with @runjerry weeks ago when this PR was submitted. The decision was to keep it the same as target update.

Reply:
> The scheduler uses global training iter as the counter, so `period` here is scheduled according to how many training iterations. But `self._counter` is maintained according to num of updates. Is this an expected behavior?

Adjusted and added comments: https://github.com/HorizonRobotics/alf/pull/1603